typha/pkg/syncserver/sync_server.go

```diff
@@ -481,6 +504,80 @@ func (s *Server) governNumberOfConnections(cxt context.Context) {
 	}
 }
 
+func (s *Server) handleGracefulShutDown(cxt context.Context, serverCancelFn context.CancelFunc) {
+	defer s.Finished.Done()
+	logCxt := log.WithField("thread", "gracefulShutdown")
+
+	select {
+	case <-s.shutdownC:
+		logCxt.Info("Graceful shutdown triggered, starting to close connections...")
+	case <-cxt.Done():
+		logCxt.Info("Context asked us to stop")
+		return
+	}
+
+	numConns := s.NumActiveConnections()
+	// Aim to close connections within 95% of the allotted time.
+	dropInterval := s.config.ShutdownTimeout * 95 / 100 / time.Duration(numConns)
+	logCtx := log.WithFields(log.Fields{
+		"activeConnections":   numConns,
+		"shutdownTimeout":     s.config.ShutdownTimeout,
+		"maximumDropInterval": s.config.ShutdownMaxDropInterval,
+	})
+	if dropInterval > s.config.ShutdownMaxDropInterval {
+		// We have a long time to shut down (say 5 minutes) but only a few connections.  Cap the delay between
+		// dropping connections.
+		dropInterval = s.config.ShutdownMaxDropInterval
+		logCtx.Info("Using maximum shutdown drop interval.")
+	} else {
+		logCxt.WithField("dropInterval", dropInterval).Info("Calculated drop interval from shutdown timeout.")
+	}
+	ticker := jitter.NewTicker(dropInterval*95/100, dropInterval*10/100)
+	for {
+		select {
+		case <-ticker.C:
+			numConns := s.NumActiveConnections()
+			logCxt := logCxt.WithField("remainingConns", numConns)
+			s.TerminateRandomConnection(logCxt, "graceful shutdown in progress")
+			if numConns <= 1 {
+				logCxt.Info("Finished closing connections, completing shut down...")
+				// Note: we release the lock between NumActiveConnections and TerminateRandomConnection so,
+				// in theory, if we haven't yet closed the listen socket, a new connection could just have been added.
+				// We don't need to worry about that because serverCancelFn will shut down all remaining connections
+				// by canceling their parent context.
+				serverCancelFn()
+				break
+			}
+		case <-cxt.Done():
+			logCxt.Info("Context asked us to stop")
+			return
+		}
+	}
+}
+
+func (s *Server) NumActiveConnections() int {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	return len(s.connIDToConn)
+}
+
+func (s *Server) ShuttingDown() bool {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	return s.shuttingDown
+}
+
+func (s *Server) TerminateRandomConnection(logCtx *log.Entry, reason string) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	for connID, conn := range s.connIDToConn {
+		logCtx.WithField("connID", connID).Infof("Closing connection; reason: %s.", reason)
+		conn.cancelCxt()
+		counterNumConnectionsDropped.Inc()
```

Comment:
Does it make sense to treat graceful termination connection drops as rebalancing? Or should we have a separate counter for them?

I'm not sure either way, but I can imagine some users wanting to distinguish between them.

Reply:
Moved this up to the caller; I think there's a good way to track shutdown already (num connections metric) so I've only kept the rebalancing one.